docs/Changelog.md

```diff
@@ -24582,6 +24582,72 @@ This version of the operator has been available since version 21 of the default
 <dd>'x_scale' determines the output type.</dd>
 </dl>
 
+### <a name="GroupNormalization-21"></a>**GroupNormalization-21**</a>
+
+  A GroupNormalization function. Carries out group normalization as described in
+  the paper https://arxiv.org/abs/1803.08494
+
+  This operator transforms input according to
+  ```
+  y = scale * (x - mean) / sqrt(variance + epsilon) + bias,
+  ```
+  where the mean and variance are computed per instance per group of channels, and
+  `scale` and `bias` should be specified for each group of channels. The number of
+  groups `num_groups` should be divisible by the number of channels so that there are
+  an equal number of channels per group.
+
+  The overall computation has two stages: the first stage normalizes the elements to
+  have zero mean and unit variance for each instance in each group, and the second
+  stage scales and shifts the results of the first stage. The floating-point precision
+  used in the first stage is determined by the `stash_type` attribute. For example,
+  if `stash_type` is 1, the operator casts all input variables to 32-bit float,
+  performs the computation, and finally casts the normalized results back to the
+  original type of `X`. The second stage does not edepend on `stash_type`.
+
+  When the number of groups is the same as the number of channels, this operator is
+  equivalent to InstanceNormalization. When there is only one group, this operator
+  is equivalent to LayerNormalization.
+
+#### Version
+
+This version of the operator has been available since version 21 of the default ONNX operator set.
+
+#### Attributes
+
+<dl>
+<dt><tt>epsilon</tt> : float (default is 1e-05)</dt>
+<dd>The epsilon value to use to avoid division by zero.</dd>
+<dt><tt>num_groups</tt> : int (required)</dt>
+<dd>The number of groups of channels. It should be a divisor of the number of channels `C`.</dd>
+<dt><tt>stash_type</tt> : int (default is 1)</dt>
+<dd>The floating-point precision used in stage one of the computation.</dd>
```

Comment:
Do we know the stash_type used in pytorch? Does it have a default type?

Reply:
PyTorch does not have a stash_type attribute. This attribute is useful for the inference setting, just like in the case of LayerNormalization.

Reply:
Right. But do you have an implicit default? Do we know what PyTorch does on the background?

Reply:
So should we allow an option, for example `-1`, to mirror PyTorch's behavior?

Reply:
I think it's a matter of taste. From a utility point of view it doesn't add anything more, since the user can simply set stash_type to match input type.

And if we do decide to make this change we should probably update the behavior of LayerNorm as well.

Reply:
When we create ONNX functions for pytorch operators, it would be good to be able to follow pytorch behavior. Otherwise the ONNX functions will need to be specialized for each invocation. @gramalingam

Reply:
I see @yuanyao-nv mentions in another comment-thread that it would be better to allow all input types as stash type. If/when we do that, it seems more natural to say that the default value of stash type is the same as input type? This would also be more consistent with the older GroupNorm op (otherwise, the version-converter would need to add the correct stash-type attribute).

I do see the point of making this consistent with LayerNorm. Perhaps that could be done as part of the other proposal to unify the various normalization functions?

Reply:
@gramalingam Are you aware of any reason LayerNorm didn't make the default `stash_type` match the input type? I wonder if there are other considerations like kernel optimization which requires the stage 1 precision to be known statically. If we make it match the input dtype then it could become a dynamically known precision.

On the other hand, IMO backward compatibility is a nice-to-have but is also a lesser concern since I'm not aware of models that use GroupNorm-18 yet due to its over-restrictiveness.

Reply:
> I wonder if there are other considerations like kernel optimization which requires the stage 1 precision to be known statically.

I don't think that is a consideration. In principle, it is still known statically since all types are known statically. I am not sure why LayerNorm had this default. There may have been no conscious reason.